src/coreclr/src/jit/gentree.cpp

```diff
@@ -12498,6 +12498,26 @@ GenTree* Compiler::gtFoldTypeCompare(GenTree* tree)
         objOp = opOther->AsCall()->gtCallThisArg->GetNode();
     }
 
+    bool                 pIsExact   = false;
+    bool                 pIsNonNull = false;
+    CORINFO_CLASS_HANDLE objCls     = gtGetClassHandle(objOp, &pIsExact, &pIsNonNull);
+
+    // if both classes are "final" (e.g. System.String[]) we can replace the comparison
+    // with `true/false` + null check.
+    if ((objCls != NO_CLASS_HANDLE) &&
+        // double check if it's final via impIsClassExact
+        impIsClassExact(objCls) && impIsClassExact(clsHnd))
+    {
+        const bool typesAreEqual = objCls == clsHnd;
+        const bool operatorIsEQ  = oper == GT_EQ;
+        const int  compareResult = operatorIsEQ ^ typesAreEqual ? 0 : 1;
+
+        // we still have to emit a null-check
+        // obj.GetType == typeof() -> (nullcheck) true/false
+        GenTree* nullcheck = gtNewNullCheck(objOp, compCurBB);
```

Comment:
And preserve any other side effects `objOp` may have.

If you are tempted to use `pIsNonNull`, then the more general form is something like:

* if `pIsNonNull` is true, and `objOp` has no side effects: `compareResult`
* if `pIsNonNull` is true, and `objOp` has side effects: `comma(objOp, compareResult)`
* if `pIsNonNull` is false, `comma(nullcheck(objOp), compareResult)`

But might not be worth the trouble, as it probably gets cleaned up later.

Reply:
I hope I addressed this correctly in my recent commit